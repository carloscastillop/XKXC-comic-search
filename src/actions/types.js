export const FETCH_COMIC = 'FETCH_COMIC';
export const FETCH_ERROR = 'FETCH_ERROR';
