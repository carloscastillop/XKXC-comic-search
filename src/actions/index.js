import {LATEST_COMIC, FETCH_COMIC, FETCH_ERROR} from './types';
import axios from 'axios';

const apiUrl = 'https://xkcd.now.sh/?comic='; //latest or 1000

export const fetchComic = (comic) => {
    return {
        type: FETCH_COMIC,
        comic
    };
};

export const latestComic = (comic) => {
    return {
        type: LATEST_COMIC,
        comic
    };
};

export const fetchError = (error) => {
    return {
        type: FETCH_ERROR,
        error
    };
};

export const fetchComics = (num = 'latest') => {
    return (dispatch) => {
        dispatch(fetchComic(null));
        dispatch(fetchError(false));
        return axios.get(apiUrl+num, {
            crossDomain: true,
            dataType: 'json',
            logLevel: 'debug',
            changeOrigin: true
        })
            .then(response => {
                dispatch(fetchComic(response.data));
                if(num === 'latest'){
                    dispatch(latestComic(response.data));
                }
            })
            .catch(error => {
                dispatch(fetchError(true));
                throw(error);
            });
    };
};
