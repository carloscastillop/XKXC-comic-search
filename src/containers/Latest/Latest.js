import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import style from '../../styles/global.module.scss';
import Comic from '../../components/Comic/Comic';
import Loading from '../../components/Loading/Loading';
import Error from '../../components/Error/Error';

function Latest({comic, error}) {

    if (error) {
        return (<Error/>);
    }

    if (!comic) {
        return (<Loading/>);
    }
    return (
        <section
            data-testid={'latest'}
            className={`${style.container} ${style['text-center']}`}
        >
            <Comic
                comic={comic}
            />
        </section>
    );
}

const mapStateToProps = state => {
    return {
        comic: state.latestComic,
        error: state.fetchError
    };
};

export default connect(
    mapStateToProps
)(Latest);

Latest.propTypes = {
    comic: PropTypes.object,
    error: PropTypes.bool
};