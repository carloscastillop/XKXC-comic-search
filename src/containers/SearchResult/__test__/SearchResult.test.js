import React from 'react';
import {render} from '@testing-library/react';
import Latest from '../SearchResult';
import {Provider} from 'react-redux';
import rootReducer from '../../../reducers';
import {applyMiddleware, createStore} from 'redux';
import thunk from 'redux-thunk';
import {fetchComic} from '../../../actions';
import {Route, MemoryRouter} from 'react-router-dom';
import renderer from 'react-test-renderer';


const comic = {
    month: '6',
    num: 2322,
    link: '',
    year: '2020',
    news: '',
    safe_title: 'ISO Paper Size Golden Spiral',
    transcript: '',
    alt: 'The ISO 216 standard ratio is cos(45Â°), but American letter paper is 8.5x11 because it uses radians, and 11/8.5 = pi/4.',
    img: 'https://imgs.xkcd.com/comics/iso_paper_size_golden_spiral.png',
    title: 'ISO Paper Size Golden Spiral',
    day: '19'
};
const store = createStore(
    rootReducer,
    applyMiddleware(thunk)
);
store.dispatch(fetchComic(comic));

test('Rendering search result', () => {
    const {getByTestId} = render(
        <Provider store={store}>
            <MemoryRouter initialEntries={['/1']}>
                <Route path='/:num'>
                    <Latest/>
                </Route>
            </MemoryRouter>
        </Provider>
    );

    const linkElement = getByTestId('search-result');
    expect(linkElement).toBeInTheDocument();
});

test('Search result comic snapshop', () => {
    const tree = renderer
        .create(
            <Provider store={store}>
                <MemoryRouter initialEntries={['/1']}>
                    <Route path='/:num'>
                        <Latest/>
                    </Route>
                </MemoryRouter>
            </Provider>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});