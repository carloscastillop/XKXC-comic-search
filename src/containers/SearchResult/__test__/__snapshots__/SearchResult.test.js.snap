// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search result comic snapshop 1`] = `
<section
  className="container text-center"
  data-testid="search-result"
>
  <div
    className="spinner"
    data-testid="loading"
  >
    <div
      className="doubleBounce1"
    />
    <div
      className="doubleBounce2"
    />
  </div>
</section>
`;
